add -race flag to the test suite

~~`project.check.sh` does some sanity checks to ensure a certain level of code quality. The go tools have a lot of useful things to provide. The question is what is worth it integrating it into the project check script. Things to consider.~~

The project check seems to be good enough for now. Anyway we need to add the `-race` flag to the test suite. Just for fun I ran the test suite using `-race`. Looks like all is good, besides the `scheduler` and `storage` packages. 

```
$ make gotest
ok  	_/home/vagrant/projects/private/anna/client/control/log	1.110s	coverage: 96.0% of statements
ok  	_/home/vagrant/projects/private/anna/client/interface/text	1.055s	coverage: 98.1% of statements
ok  	_/home/vagrant/projects/private/anna/factory/id	1.408s	coverage: 96.3% of statements
ok  	_/home/vagrant/projects/private/anna/factory/permutation	1.017s	coverage: 100.0% of statements
ok  	_/home/vagrant/projects/private/anna/factory/random	1.039s	coverage: 100.0% of statements
ok  	_/home/vagrant/projects/private/anna/gateway	1.028s	coverage: 93.3% of statements
ok  	_/home/vagrant/projects/private/anna/index/clg	1.672s	coverage: 67.7% of statements
ok  	_/home/vagrant/projects/private/anna/index/clg/collection	9.417s	coverage: 98.0% of statements
ok  	_/home/vagrant/projects/private/anna/index/clg/collection/distribution	1.042s	coverage: 85.7% of statements
ok  	_/home/vagrant/projects/private/anna/index/clg/collection/feature-set	2.224s	coverage: 96.7% of statements
ok  	_/home/vagrant/projects/private/anna/index/clg/profile	1.659s	coverage: 65.9% of statements
ok  	_/home/vagrant/projects/private/anna/log	1.032s	coverage: 97.2% of statements
ok  	_/home/vagrant/projects/private/anna/net/pat	1.025s	coverage: 78.9% of statements
==================
WARNING: DATA RACE
Write by goroutine 55:
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Boot_Reschedule.func1()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:65 +0x7b
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).waitForAction()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:124 +0x223
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func3()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:179 +0xa0

Previous read by goroutine 52:
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Boot_Reschedule.func1()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:67 +0xa8
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).waitForAction()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:124 +0x223
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func3()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:179 +0xa0

Goroutine 55 (running) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:184 +0x6e9
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).scheduleActiveJobs.func1()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:27 +0x17a
  github.com/xh3b4sd/anna/storage/memory.(*storage).WalkScoredElements()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:323 +0x378
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).scheduleActiveJobs()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:35 +0x2ed
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Boot.func1.1()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:126 +0x4a

Goroutine 52 (running) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:184 +0x6e9
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).scheduleActiveJobs.func1()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:27 +0x17a
  github.com/xh3b4sd/anna/storage/memory.(*storage).WalkScoredElements()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:323 +0x378
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).scheduleActiveJobs()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:35 +0x2ed
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Boot.func1.1()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:126 +0x4a
==================
==================
WARNING: DATA RACE
Write by goroutine 74:
  runtime.mapdelete()
      /home/vagrant/.gvm/gos/go1.6/src/runtime/hashmap.go:545 +0x0
  github.com/xh3b4sd/anna/storage/memory.(*storage).RemoveScoredElement()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:213 +0x337
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).MarkAsInactive()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:277 +0x606
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).MarkAsSucceeded()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:315 +0x23b
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).waitForAction()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:156 +0x776
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func3()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:179 +0xa0

Previous read by goroutine 72:
  runtime.mapaccess2()
      /home/vagrant/.gvm/gos/go1.6/src/runtime/hashmap.go:329 +0x0
  github.com/xh3b4sd/anna/storage/memory.(*storage).GetElementsByScore()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:110 +0x1e7
  github.com/xh3b4sd/anna/storage/memory.(*storage).GetHighestScoredElements()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:149 +0x2e3
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).preventDuplicates()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:46 +0x229
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func1()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:159 +0x79

Goroutine 74 (running) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:184 +0x6e9
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Success()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:147 +0x10a
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc

Goroutine 72 (finished) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:164 +0x5fe
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Success()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:147 +0x10a
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc
==================
==================
WARNING: DATA RACE
Write by goroutine 79:
  runtime.mapdelete()
      /home/vagrant/.gvm/gos/go1.6/src/runtime/hashmap.go:545 +0x0
  github.com/xh3b4sd/anna/storage/memory.(*storage).RemoveScoredElement()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:213 +0x337
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).MarkAsInactive()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:277 +0x606
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).MarkAsFailedWithError()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:248 +0x26c
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).waitForAction()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:127 +0x2bf
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func3()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:179 +0xa0

Previous read by goroutine 28:
  runtime.mapaccess2()
      /home/vagrant/.gvm/gos/go1.6/src/runtime/hashmap.go:329 +0x0
  github.com/xh3b4sd/anna/storage/memory.(*storage).GetElementsByScore()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:110 +0x1e7
  github.com/xh3b4sd/anna/storage/memory.(*storage).GetHighestScoredElements()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:149 +0x2e3
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).preventDuplicates()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:46 +0x229
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func1()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:159 +0x79

Goroutine 79 (running) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:184 +0x6e9
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Action_Error()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:208 +0x10a
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc

Goroutine 28 (finished) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:164 +0x5fe
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Action_Error()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:208 +0x10a
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc
==================
[16/05/28 21:15:35.628] [L: E] [O: scheduler / 002f0530c56c0253] [V:  4] [{_/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:182: } {_/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:133: } {test error}]
==================
WARNING: DATA RACE
Write by goroutine 54:
  runtime.mapdelete()
      /home/vagrant/.gvm/gos/go1.6/src/runtime/hashmap.go:545 +0x0
  github.com/xh3b4sd/anna/storage/memory.(*storage).RemoveScoredElement()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:213 +0x337
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).MarkAsInactive()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:277 +0x606
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).MarkAsSucceeded()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:315 +0x23b
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Scheduler_Ends_Wait()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:339 +0x45c
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc

Previous read by goroutine 68:
  runtime.mapaccess2()
      /home/vagrant/.gvm/gos/go1.6/src/runtime/hashmap.go:329 +0x0
  github.com/xh3b4sd/anna/storage/memory.(*storage).GetElementsByScore()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:110 +0x1e7
  github.com/xh3b4sd/anna/storage/memory.(*storage).GetHighestScoredElements()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:149 +0x2e3
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).preventDuplicates()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:46 +0x229
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func1()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:159 +0x79

Goroutine 54 (running) created at:
  testing.RunTests()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:582 +0xae2
  testing.(*M).Run()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:515 +0x11d
  main.main()
      _/home/vagrant/projects/private/anna/scheduler/_test/_testmain.go:154 +0x385

Goroutine 68 (finished) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:164 +0x5fe
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Scheduler_Ends_Wait()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:332 +0x17a
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc
==================
==================
WARNING: DATA RACE
Write by goroutine 70:
  sync/atomic.CompareAndSwapInt32()
      /home/vagrant/.gvm/gos/go1.6/src/runtime/race_amd64.s:279 +0xb
  sync.(*Mutex).Lock()
      /home/vagrant/.gvm/gos/go1.6/src/sync/mutex.go:44 +0x4d
  _/home/vagrant/projects/private/anna/scheduler.(*job).GetArgs()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/job.go:149 +0x6d
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).waitForAction()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:124 +0x1e6
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func3()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:179 +0xa0

Previous read by goroutine 54:
  _/home/vagrant/projects/private/anna/scheduler.(*job).MarshalJSON()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/job_json.go:17 +0x90
  encoding/json.marshalerEncoder()
      /home/vagrant/.gvm/gos/go1.6/src/encoding/json/encode.go:418 +0x16e
  encoding/json.(*encodeState).reflectValue()
      /home/vagrant/.gvm/gos/go1.6/src/encoding/json/encode.go:301 +0x78
  encoding/json.(*encodeState).marshal()
      /home/vagrant/.gvm/gos/go1.6/src/encoding/json/encode.go:274 +0xb6
  encoding/json.Marshal()
      /home/vagrant/.gvm/gos/go1.6/src/encoding/json/encode.go:139 +0xbf
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).StoreJob()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:335 +0x207
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).MarkAsSucceeded()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:321 +0x365
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Scheduler_Ends_Wait()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:339 +0x45c
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc

Goroutine 70 (running) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:184 +0x6e9
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Scheduler_Ends_Wait()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:332 +0x17a
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc

Goroutine 54 (running) created at:
  testing.RunTests()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:582 +0xae2
  testing.(*M).Run()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:515 +0x11d
  main.main()
      _/home/vagrant/projects/private/anna/scheduler/_test/_testmain.go:154 +0x385
==================
PASS
coverage: 90.3% of statements
Found 5 data race(s)
FAIL	_/home/vagrant/projects/private/anna/scheduler	2.262s
make: *** [gotest] Error 1
```

So even if the test suite will take much longer to complete, we need to add the `-race` flag to the test suite and fix the found issues. 

See
- https://blog.golang.org/race-detector
- https://golang.org/doc/articles/race_detector.html

Just for fun I ran the test suite using `-race`. Looks like all is good, besides the `scheduler` and `storage` packages. 

```
$ make gotest
ok  	_/home/vagrant/projects/private/anna/client/control/log	1.110s	coverage: 96.0% of statements
ok  	_/home/vagrant/projects/private/anna/client/interface/text	1.055s	coverage: 98.1% of statements
ok  	_/home/vagrant/projects/private/anna/factory/id	1.408s	coverage: 96.3% of statements
ok  	_/home/vagrant/projects/private/anna/factory/permutation	1.017s	coverage: 100.0% of statements
ok  	_/home/vagrant/projects/private/anna/factory/random	1.039s	coverage: 100.0% of statements
ok  	_/home/vagrant/projects/private/anna/gateway	1.028s	coverage: 93.3% of statements
ok  	_/home/vagrant/projects/private/anna/index/clg	1.672s	coverage: 67.7% of statements
ok  	_/home/vagrant/projects/private/anna/index/clg/collection	9.417s	coverage: 98.0% of statements
ok  	_/home/vagrant/projects/private/anna/index/clg/collection/distribution	1.042s	coverage: 85.7% of statements
ok  	_/home/vagrant/projects/private/anna/index/clg/collection/feature-set	2.224s	coverage: 96.7% of statements
ok  	_/home/vagrant/projects/private/anna/index/clg/profile	1.659s	coverage: 65.9% of statements
ok  	_/home/vagrant/projects/private/anna/log	1.032s	coverage: 97.2% of statements
ok  	_/home/vagrant/projects/private/anna/net/pat	1.025s	coverage: 78.9% of statements
==================
WARNING: DATA RACE
Write by goroutine 55:
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Boot_Reschedule.func1()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:65 +0x7b
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).waitForAction()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:124 +0x223
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func3()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:179 +0xa0

Previous read by goroutine 52:
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Boot_Reschedule.func1()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:67 +0xa8
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).waitForAction()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:124 +0x223
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func3()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:179 +0xa0

Goroutine 55 (running) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:184 +0x6e9
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).scheduleActiveJobs.func1()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:27 +0x17a
  github.com/xh3b4sd/anna/storage/memory.(*storage).WalkScoredElements()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:323 +0x378
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).scheduleActiveJobs()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:35 +0x2ed
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Boot.func1.1()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:126 +0x4a

Goroutine 52 (running) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:184 +0x6e9
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).scheduleActiveJobs.func1()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:27 +0x17a
  github.com/xh3b4sd/anna/storage/memory.(*storage).WalkScoredElements()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:323 +0x378
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).scheduleActiveJobs()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:35 +0x2ed
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Boot.func1.1()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:126 +0x4a
==================
==================
WARNING: DATA RACE
Write by goroutine 74:
  runtime.mapdelete()
      /home/vagrant/.gvm/gos/go1.6/src/runtime/hashmap.go:545 +0x0
  github.com/xh3b4sd/anna/storage/memory.(*storage).RemoveScoredElement()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:213 +0x337
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).MarkAsInactive()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:277 +0x606
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).MarkAsSucceeded()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:315 +0x23b
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).waitForAction()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:156 +0x776
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func3()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:179 +0xa0

Previous read by goroutine 72:
  runtime.mapaccess2()
      /home/vagrant/.gvm/gos/go1.6/src/runtime/hashmap.go:329 +0x0
  github.com/xh3b4sd/anna/storage/memory.(*storage).GetElementsByScore()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:110 +0x1e7
  github.com/xh3b4sd/anna/storage/memory.(*storage).GetHighestScoredElements()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:149 +0x2e3
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).preventDuplicates()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:46 +0x229
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func1()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:159 +0x79

Goroutine 74 (running) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:184 +0x6e9
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Success()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:147 +0x10a
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc

Goroutine 72 (finished) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:164 +0x5fe
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Success()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:147 +0x10a
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc
==================
==================
WARNING: DATA RACE
Write by goroutine 79:
  runtime.mapdelete()
      /home/vagrant/.gvm/gos/go1.6/src/runtime/hashmap.go:545 +0x0
  github.com/xh3b4sd/anna/storage/memory.(*storage).RemoveScoredElement()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:213 +0x337
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).MarkAsInactive()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:277 +0x606
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).MarkAsFailedWithError()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:248 +0x26c
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).waitForAction()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:127 +0x2bf
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func3()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:179 +0xa0

Previous read by goroutine 28:
  runtime.mapaccess2()
      /home/vagrant/.gvm/gos/go1.6/src/runtime/hashmap.go:329 +0x0
  github.com/xh3b4sd/anna/storage/memory.(*storage).GetElementsByScore()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:110 +0x1e7
  github.com/xh3b4sd/anna/storage/memory.(*storage).GetHighestScoredElements()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:149 +0x2e3
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).preventDuplicates()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:46 +0x229
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func1()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:159 +0x79

Goroutine 79 (running) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:184 +0x6e9
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Action_Error()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:208 +0x10a
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc

Goroutine 28 (finished) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:164 +0x5fe
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Action_Error()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:208 +0x10a
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc
==================
[16/05/28 21:15:35.628] [L: E] [O: scheduler / 002f0530c56c0253] [V:  4] [{_/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:182: } {_/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:133: } {test error}]
==================
WARNING: DATA RACE
Write by goroutine 54:
  runtime.mapdelete()
      /home/vagrant/.gvm/gos/go1.6/src/runtime/hashmap.go:545 +0x0
  github.com/xh3b4sd/anna/storage/memory.(*storage).RemoveScoredElement()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:213 +0x337
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).MarkAsInactive()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:277 +0x606
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).MarkAsSucceeded()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:315 +0x23b
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Scheduler_Ends_Wait()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:339 +0x45c
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc

Previous read by goroutine 68:
  runtime.mapaccess2()
      /home/vagrant/.gvm/gos/go1.6/src/runtime/hashmap.go:329 +0x0
  github.com/xh3b4sd/anna/storage/memory.(*storage).GetElementsByScore()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:110 +0x1e7
  github.com/xh3b4sd/anna/storage/memory.(*storage).GetHighestScoredElements()
      /home/vagrant/projects/private/anna/.workspace/src/github.com/xh3b4sd/anna/storage/memory/memory_storage.go:149 +0x2e3
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).preventDuplicates()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:46 +0x229
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func1()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:159 +0x79

Goroutine 54 (running) created at:
  testing.RunTests()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:582 +0xae2
  testing.(*M).Run()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:515 +0x11d
  main.main()
      _/home/vagrant/projects/private/anna/scheduler/_test/_testmain.go:154 +0x385

Goroutine 68 (finished) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:164 +0x5fe
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Scheduler_Ends_Wait()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:332 +0x17a
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc
==================
==================
WARNING: DATA RACE
Write by goroutine 70:
  sync/atomic.CompareAndSwapInt32()
      /home/vagrant/.gvm/gos/go1.6/src/runtime/race_amd64.s:279 +0xb
  sync.(*Mutex).Lock()
      /home/vagrant/.gvm/gos/go1.6/src/sync/mutex.go:44 +0x4d
  _/home/vagrant/projects/private/anna/scheduler.(*job).GetArgs()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/job.go:149 +0x6d
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).waitForAction()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/common.go:124 +0x1e6
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute.func3()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:179 +0xa0

Previous read by goroutine 54:
  _/home/vagrant/projects/private/anna/scheduler.(*job).MarshalJSON()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/job_json.go:17 +0x90
  encoding/json.marshalerEncoder()
      /home/vagrant/.gvm/gos/go1.6/src/encoding/json/encode.go:418 +0x16e
  encoding/json.(*encodeState).reflectValue()
      /home/vagrant/.gvm/gos/go1.6/src/encoding/json/encode.go:301 +0x78
  encoding/json.(*encodeState).marshal()
      /home/vagrant/.gvm/gos/go1.6/src/encoding/json/encode.go:274 +0xb6
  encoding/json.Marshal()
      /home/vagrant/.gvm/gos/go1.6/src/encoding/json/encode.go:139 +0xbf
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).StoreJob()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:335 +0x207
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).MarkAsSucceeded()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:321 +0x365
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Scheduler_Ends_Wait()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:339 +0x45c
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc

Goroutine 70 (running) created at:
  _/home/vagrant/projects/private/anna/scheduler.(*scheduler).Execute()
      _/home/vagrant/projects/private/anna/scheduler/_test/_obj_test/scheduler.go:184 +0x6e9
  _/home/vagrant/projects/private/anna/scheduler.Test_Scheduler_Execute_Scheduler_Ends_Wait()
      /home/vagrant/projects/private/anna/scheduler/scheduler_test.go:332 +0x17a
  testing.tRunner()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:473 +0xdc

Goroutine 54 (running) created at:
  testing.RunTests()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:582 +0xae2
  testing.(*M).Run()
      /home/vagrant/.gvm/gos/go1.6/src/testing/testing.go:515 +0x11d
  main.main()
      _/home/vagrant/projects/private/anna/scheduler/_test/_testmain.go:154 +0x385
==================
PASS
coverage: 90.3% of statements
Found 5 data race(s)
FAIL	_/home/vagrant/projects/private/anna/scheduler	2.262s
make: *** [gotest] Error 1
```

So even the test suite will take much longer to complete, we need to add the `-race` flag to the test suite and fix the found issues.

--- a/doc/development/cheat_sheet.md
+++ b/doc/development/cheat_sheet.md
@@ -37,9 +37,11 @@ find . -not -path "./.git/*" -not -path "./.workspace/*" | xargs misspell -w
 The following snippet can be used to easily run unit tests of single packages.
 This is unlike to the makefile target, that runs the tests of all packages.
 Here all tests of the package `pkg` will be run. For more options consider
-reading `go test -h`.
+reading `go test -h`. Note that it is a good idea to run tests with having the
+race detector enabled. For more information on this one see
+https://blog.golang.org/race-detector.
 ```
-GOPATH=$(pwd)/.workspace/ go test ./pkg
+GOPATH=$(pwd)/.workspace/ go test -race ./pkg
 ```
 
 ### search and replace

--- a/go.test.sh
+++ b/go.test.sh
@@ -8,7 +8,7 @@ set -e
 > coverage.txt
 
 for d in $(find ./* -maxdepth 10 -type f -name '*test.go' -not -path "./.workspace/*" -not -path "./.git/*" -exec dirname {} \; | sort -u); do
-  go test -covermode=atomic -coverprofile=profile.out $d
+  go test -race -covermode=atomic -coverprofile=profile.out $d
   if [ -f profile.out ]; then
       cat profile.out >> coverage.txt
       rm profile.out

